File: C++/busca_binaria.cpp
#include <iostream>

void burble_sort(int tamanho, int vetor[]){
    int auxiliar;
    for(int j = 0; j < tamanho; j++){
        for(int i = 0; i < tamanho - 1; i++){
            if(vetor[i] > vetor[i+1]){
                auxiliar = vetor[i];
                vetor[i] = vetor[i+1];
                vetor[i+1] = auxiliar;
            }
        }
    }
}

void busca_binaria(int numero, int vetor[], int inicio, int fim){
    bool encontrou = false;
    int meio;

    while (encontrou == false){
        meio = inicio + (fim - inicio) / 2;

        if(numero == vetor[meio]){
            encontrou = true;
        }
        else if(numero < vetor[meio]){
            fim = meio - 1;
        }
        else if(numero > vetor[meio]){
            inicio = meio + 1;
        }
        else{
            break;
        }
    }
    if(encontrou == false){
        std::cout << "Número não encontrado no vetor." << std::endl;
    }
    else{
        std::cout << "O valor " << numero << " está na posição " << meio << ", onde o vetor se inicia com o número " << vetor[0] << " na posição 0." << std::endl;
    }
}

int main(){
    int vetor[] = {30, 28, 83, 14, 56, 356, 97, 108, 9999, 200}, numero, inicio = 0,fim = sizeof(vetor) / sizeof(vetor[0]);

    burble_sort(fim, vetor);

    std::cout << "Escolha o número que deseja buscar no vetor [";
    for(int i = 0; i < fim; i++){
        std::cout << vetor[i];
        if (i < fim - 1){
            std::cout << ", ";
        }
    }
    std::cout << "]: ";
    std:: cin >> numero;

    busca_binaria(numero, vetor, inicio, fim);

    return 0;
}

File: python/busca_binaria.py

def burble_sort(vetor):
    for i in range(len(vetor)):
        for j in range(len(vetor) - 1):
            if(vetor[j] > vetor[j+1]):
                auxiliar = vetor[j]
                vetor[j] = vetor[j+1]
                vetor[j+1] = auxiliar
    
def busca_binaria(numero, vetor):
    encontrado  = False
    inicio = 0
    fim = len(vetor) - 1 
    
    while inicio <= fim and encontrado == False:
        meio = (fim + inicio) // 2
        
        if numero == vetor[meio]:
            encontrado = True
        elif numero < vetor[meio]:
            fim = meio - 1
        else:
            inicio = meio + 1
        
    if encontrado:
        print(f"O número {numero} está na posição {meio} do vetor, onde o vetor se inicia com o número {vetor[0]} na posição 0.")
    else:
        print(f"O número {numero} não está presente no vetor.")


vetor = [5, 1, 39, 28, 104, 90, 80, -39]

burble_sort(vetor)

numero = int(input(f"Digite o número que deseja buscar no vetor {vetor}: "))

busca_binaria(numero, vetor)
